File: src/lsp/semantic_tokens.rs
use crate::ast::{Expr, Stmt};
use crate::lsp::store::DocumentStore;
use lsp_types::*;

// Legend: token type index
// 0 namespace, 1 type(enum/struct), 2 class, 3 function, 4 property, 5 method, 6 variable,
// 7 parameter, 8 keyword, 9 string, 10 number, 11 comment, 12 operator, 13 decorator
const TYPE_KEYWORD: u32 = 8;
const TYPE_VARIABLE: u32 = 6;
const TYPE_FUNCTION: u32 = 3;
const TYPE_METHOD: u32 = 5;
const TYPE_CLASS: u32 = 2;
const TYPE_PARAMETER: u32 = 7;
const TYPE_STRING: u32 = 9;
const TYPE_NUMBER: u32 = 10;
const TYPE_COMMENT: u32 = 11;
const TYPE_DECORATOR: u32 = 13;
const TYPE_ENUM: u32 = 1;
const TYPE_STRUCT: u32 = 1;

pub fn handle_semantic_tokens(
    store: &mut DocumentStore,
    params: SemanticTokensParams,
) -> Option<SemanticTokensResult> {
    store.reparse_if_dirty(&params.text_document.uri);
    let doc = store.get(&params.text_document.uri)?;
    let tokens = extract_semantic_tokens(&doc.ast, &doc.source);

    // Sort by position for proper delta encoding
    let mut tokens = tokens;
    tokens.sort_by_key(|t| (t.line, t.character));

    let data = encode_semantic_tokens(&tokens);
    Some(SemanticTokensResult::Tokens(SemanticTokens {
        result_id: None,
        data,
    }))
}

#[derive(Debug, Clone)]
struct InternalToken {
    line: u32,
    character: u32,
    length: u32,
    token_type: u32,
    modifiers: u32,
}

fn encode_semantic_tokens(tokens: &[InternalToken]) -> Vec<SemanticToken> {
    let mut encoded = Vec::with_capacity(tokens.len());
    let mut prev_line: u32 = 0;
    let mut prev_char: u32 = 0;

    for t in tokens {
        let delta_line = t.line - prev_line;
        let delta_char = if delta_line == 0 {
            t.character - prev_char
        } else {
            t.character
        };
        encoded.push(SemanticToken {
            delta_line,
            delta_start: delta_char,
            length: t.length,
            token_type: t.token_type,
            token_modifiers_bitset: t.modifiers,
        });
        prev_line = t.line;
        prev_char = t.character;
    }
    encoded
}

fn extract_semantic_tokens(ast: &[Stmt], source: &str) -> Vec<InternalToken> {
    let mut tokens = Vec::new();
    walk_stmts(ast, source, &mut tokens, false);
    // Also extract comment tokens
    tokens.extend(extract_comments(source));
    tokens
}

fn extract_comments(source: &str) -> Vec<InternalToken> {
    let mut tokens = Vec::new();
    let lines: Vec<&str> = source.lines().collect();
    for (i, line) in lines.iter().enumerate() {
        let trimmed = line.trim_start();
        if trimmed.starts_with("#") {
            let col = line.find('#').unwrap_or(0);
            tokens.push(InternalToken {
                line: i as u32,
                character: col as u32,
                length: trimmed.len() as u32,
                token_type: TYPE_COMMENT,
                modifiers: 0,
            });
        }
    }
    tokens
}

fn walk_stmts(stmts: &[Stmt], source: &str, tokens: &mut Vec<InternalToken>, inside_class: bool) {
    for stmt in stmts {
        walk_stmt(stmt, source, tokens, inside_class);
    }
}

fn walk_stmt(stmt: &Stmt, source: &str, tokens: &mut Vec<InternalToken>, inside_class: bool) {
    match stmt {
        Stmt::Local { name, value, span, .. } => {
            // Variable name
            if let Some(pos) = find_token_in_source(source, name, span.start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: name.len() as u32,
                    token_type: TYPE_VARIABLE,
                    modifiers: 0,
                });
            }
            if let Some(val) = value {
                walk_expr(val, source, span.start, tokens);
            }
        }
        Stmt::Assign { target, value, span, .. } => {
            walk_expr(target, source, span.start, tokens);
            walk_expr(value, source, span.start, tokens);
        }
        Stmt::Return { value, span, .. } => {
            tokens.push(make_keyword_token(source, span, "return"));
            if let Some(val) = value {
                walk_expr(val, source, span.start, tokens);
            }
        }
        Stmt::If { cond, then_block, else_if_blocks, else_block, span } => {
            tokens.push(make_keyword_token(source, span, "if"));
            walk_expr(cond, source, span.start, tokens);
            walk_stmts(then_block, source, tokens, inside_class);
            for (_cond, block) in else_if_blocks {
                walk_stmts(block, source, tokens, inside_class);
            }
            if let Some(block) = else_block {
                walk_stmts(block, source, tokens, inside_class);
            }
        }
        Stmt::While { cond, block, span } => {
            tokens.push(make_keyword_token(source, span, "while"));
            walk_expr(cond, source, span.start, tokens);
            walk_stmts(block, source, tokens, inside_class);
        }
        Stmt::For { var, iter, block, span } => {
            tokens.push(make_keyword_token(source, span, "for"));
            if let Some(pos) = find_token_in_source(source, var, span.start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: var.len() as u32,
                    token_type: TYPE_VARIABLE,
                    modifiers: 0,
                });
            }
            walk_expr(iter, source, span.start, tokens);
            walk_stmts(block, source, tokens, inside_class);
        }
        Stmt::FuncDef { name, params, param_types: _, param_defaults, block, access, is_async, span } => {
            // "public/private" keyword
            if access == "public" || access == "private" {
                tokens.push(make_keyword_token(source, span, access));
            }
            // "def" keyword
            tokens.push(make_keyword_token(source, span, "def"));
            if *is_async {
                tokens.push(make_keyword_token(source, span, "async"));
            }
            // Function name
            if let Some(pos) = find_token_in_source(source, name, span.start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: name.len() as u32,
                    token_type: if inside_class { TYPE_METHOD } else { TYPE_FUNCTION },
                    modifiers: 0,
                });
            }
            // Parameters
            for (i, param) in params.iter().enumerate() {
                if let Some(pos) = find_token_in_source(source, param, span.start) {
                    tokens.push(InternalToken {
                        line: pos.line,
                        character: pos.col,
                        length: param.len() as u32,
                        token_type: TYPE_PARAMETER,
                        modifiers: 0,
                    });
                }
                if let Some(default) = param_defaults.get(i).and_then(|d| d.as_ref()) {
                    walk_expr(default, source, span.start, tokens);
                }
            }
            walk_stmts(block, source, tokens, false);
        }
        Stmt::ClassDef { name, body, access, span } => {
            if access == "public" || access == "private" {
                tokens.push(make_keyword_token(source, span, access));
            }
            tokens.push(make_keyword_token(source, span, "class"));
            if let Some(pos) = find_token_in_source(source, name, span.start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: name.len() as u32,
                    token_type: TYPE_CLASS,
                    modifiers: 0,
                });
            }
            walk_stmts(body, source, tokens, true);
        }
        Stmt::ExprStmt { expr, span } => {
            walk_expr(expr, source, span.start, tokens);
        }
        Stmt::EnumDef { name, variants, access, span } => {
            if access == "public" || access == "private" {
                tokens.push(make_keyword_token(source, span, access));
            }
            tokens.push(make_keyword_token(source, span, "enum"));
            if let Some(pos) = find_token_in_source(source, name, span.start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: name.len() as u32,
                    token_type: TYPE_ENUM,
                    modifiers: 0,
                });
            }
            for variant in variants {
                if let Some(pos) = find_token_in_source(source, variant, span.start) {
                    tokens.push(InternalToken {
                        line: pos.line,
                        character: pos.col,
                        length: variant.len() as u32,
                        token_type: TYPE_VARIABLE,
                        modifiers: 0,
                    });
                }
            }
        }
        Stmt::StructDef { name, fields, access, span } => {
            if access == "public" || access == "private" {
                tokens.push(make_keyword_token(source, span, access));
            }
            tokens.push(make_keyword_token(source, span, "struct"));
            if let Some(pos) = find_token_in_source(source, name, span.start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: name.len() as u32,
                    token_type: TYPE_STRUCT,
                    modifiers: 0,
                });
            }
            for field in fields {
                if let Some(pos) = find_token_in_source(source, &field.name, span.start) {
                    tokens.push(InternalToken {
                        line: pos.line,
                        character: pos.col,
                        length: field.name.len() as u32,
                        token_type: TYPE_VARIABLE,
                        modifiers: 0,
                    });
                }
            }
        }
        Stmt::Import { path, alias, span } => {
            tokens.push(make_keyword_token(source, span, "import"));
            let token_name = if alias.is_empty() { path } else { alias };
            if let Some(pos) = find_token_in_source(source, token_name, span.start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: token_name.len() as u32,
                    token_type: TYPE_VARIABLE,
                    modifiers: 0,
                });
            }
        }
        Stmt::Break { span } => {
            tokens.push(make_keyword_token(source, span, "break"));
        }
        Stmt::Continue { span } => {
            tokens.push(make_keyword_token(source, span, "continue"));
        }
        Stmt::TryCatch { try_block, catch_clauses, finally_block, span } => {
            tokens.push(make_keyword_token(source, span, "try"));
            walk_stmts(try_block, source, tokens, inside_class);
            for (_var, _ty, block) in catch_clauses {
                tokens.push(make_keyword_token(source, span, "except"));
                walk_stmts(block, source, tokens, inside_class);
            }
            if let Some(block) = finally_block {
                tokens.push(make_keyword_token(source, span, "finally"));
                walk_stmts(block, source, tokens, inside_class);
            }
        }
        Stmt::DecoratedStmt { decorators, stmt, span } => {
            for decorator in decorators {
                if let Some(pos) = find_token_in_source(source, &format!("@{}", decorator), span.start) {
                    tokens.push(InternalToken {
                        line: pos.line,
                        character: pos.col,
                        length: (decorator.len() + 1) as u32,
                        token_type: TYPE_DECORATOR,
                        modifiers: 0,
                    });
                }
            }
            walk_stmt(stmt, source, tokens, inside_class);
        }
    }
}

fn walk_expr(expr: &Expr, source: &str, search_start: usize, tokens: &mut Vec<InternalToken>) {
    match expr {
        Expr::Number(_) => {
            if let Some(pos) = find_numeric_in_source(source, search_start) {
                let num_len = source[search_start..]
                    .find(|c: char| !c.is_ascii_digit() && c != '.')
                    .unwrap_or(source.len() - search_start);
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: num_len as u32,
                    token_type: TYPE_NUMBER,
                    modifiers: 0,
                });
            }
        }
        Expr::Str(s) => {
            let search_str = format!("{:?}", s);
            if let Some(pos) = find_token_in_source(source, &search_str, search_start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: search_str.len() as u32,
                    token_type: TYPE_STRING,
                    modifiers: 0,
                });
            }
        }
        Expr::FString(_) => {
            if let Some(pos) = find_fstring_in_source(source, search_start) {
                let end = source[search_start..]
                    .find('"')
                    .map(|e| e + 1)
                    .unwrap_or(10);
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: end as u32,
                    token_type: TYPE_STRING,
                    modifiers: 0,
                });
            }
        }
        Expr::Ident(name) => {
            if let Some(pos) = find_token_in_source(source, name, search_start) {
                // Skip keywords
                if !is_keyword(name) {
                    tokens.push(InternalToken {
                        line: pos.line,
                        character: pos.col,
                        length: name.len() as u32,
                        token_type: TYPE_VARIABLE,
                        modifiers: 0,
                    });
                }
            }
        }
        Expr::Call { func, args } => {
            if let Some(pos) = find_token_in_source(source, func, search_start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: func.len() as u32,
                    token_type: TYPE_FUNCTION,
                    modifiers: 0,
                });
            }
            for arg in args {
                walk_expr(arg, source, search_start, tokens);
            }
        }
        Expr::MethodCall { obj, field, args, .. } => {
            walk_expr(obj, source, search_start, tokens);
            if let Some(pos) = find_token_in_source(source, field, search_start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: field.len() as u32,
                    token_type: TYPE_METHOD,
                    modifiers: 0,
                });
            }
            for arg in args {
                walk_expr(arg, source, search_start, tokens);
            }
        }
        Expr::Member { obj, field, .. } => {
            walk_expr(obj, source, search_start, tokens);
            if let Some(pos) = find_token_in_source(source, field, search_start) {
                tokens.push(InternalToken {
                    line: pos.line,
                    character: pos.col,
                    length: field.len() as u32,
                    token_type: TYPE_METHOD,
                    modifiers: 0,
                });
            }
        }
        Expr::Index { obj, index } => {
            walk_expr(obj, source, search_start, tokens);
            walk_expr(index, source, search_start, tokens);
        }
        Expr::Binary { left, right, .. } => {
            walk_expr(left, source, search_start, tokens);
            walk_expr(right, source, search_start, tokens);
        }
        Expr::Logical { left, right, .. } => {
            walk_expr(left, source, search_start, tokens);
            walk_expr(right, source, search_start, tokens);
        }
        Expr::UnaryMinus(inner) | Expr::Not(inner) | Expr::Grouping(inner) => {
            walk_expr(inner, source, search_start, tokens);
        }
        Expr::Ternary { cond, then_expr, else_expr } => {
            walk_expr(cond, source, search_start, tokens);
            walk_expr(then_expr, source, search_start, tokens);
            walk_expr(else_expr, source, search_start, tokens);
        }
        Expr::Array(elems) => {
            for elem in elems {
                walk_expr(elem, source, search_start, tokens);
            }
        }
        Expr::Table(fields) => {
            for field in fields {
                match field {
                    crate::ast::TableField::Pair { key, value } => {
                        walk_expr(key, source, search_start, tokens);
                        walk_expr(value, source, search_start, tokens);
                    }
                    crate::ast::TableField::Value(v) => {
                        walk_expr(v, source, search_start, tokens);
                    }
                }
            }
        }
        Expr::Function { params, block } => {
            for param in params {
                if let Some(pos) = find_token_in_source(source, param, search_start) {
                    tokens.push(InternalToken {
                        line: pos.line,
                        character: pos.col,
                        length: param.len() as u32,
                        token_type: TYPE_PARAMETER,
                        modifiers: 0,
                    });
                }
            }
            walk_stmts(block, source, tokens, false);
        }
        Expr::ListComp { elt, generators } => {
            walk_expr(elt, source, search_start, tokens);
            for gen in generators {
                if let Some(pos) = find_token_in_source(source, &gen.var, search_start) {
                    tokens.push(InternalToken {
                        line: pos.line,
                        character: pos.col,
                        length: gen.var.len() as u32,
                        token_type: TYPE_VARIABLE,
                        modifiers: 0,
                    });
                }
                walk_expr(&gen.iter, source, search_start, tokens);
                if let Some(cond) = &gen.condition {
                    walk_expr(cond, source, search_start, tokens);
                }
            }
        }
        _ => {} // Bool, Nil, SelfExpr, AwaitExpr, etc. — no tokens needed
    }
}

#[derive(Debug, Clone, Copy)]
struct TokenPosition {
    line: u32,
    col: u32,
}

fn find_token_in_source(source: &str, token: &str, search_start: usize) -> Option<TokenPosition> {
    // Search within a window around span start
    let window = &source[search_start.min(source.len())..];
    let pos_in_window = window.find(token)?;
    let abs_pos = search_start + pos_in_window;
    byte_pos_to_line_col(source, abs_pos)
}

fn find_numeric_in_source(source: &str, search_start: usize) -> Option<TokenPosition> {
    let window = &source[search_start.min(source.len())..];
    let pos_in_window = window.find(|c: char| c.is_ascii_digit())?;
    let abs_pos = search_start + pos_in_window;
    byte_pos_to_line_col(source, abs_pos)
}

fn find_fstring_in_source(source: &str, search_start: usize) -> Option<TokenPosition> {
    let window = &source[search_start.min(source.len())..];
    let pos_in_window = window.find("f\"")?;
    let abs_pos = search_start + pos_in_window;
    byte_pos_to_line_col(source, abs_pos)
}

fn make_keyword_token(source: &str, span: &crate::ast::Span, keyword: &str) -> InternalToken {
    let pos = find_token_in_source(source, keyword, span.start)
        .unwrap_or(TokenPosition { line: 0, col: 0 });
    InternalToken {
        line: pos.line,
        character: pos.col,
        length: keyword.len() as u32,
        token_type: TYPE_KEYWORD,
        modifiers: 0,
    }
}

fn byte_pos_to_line_col(source: &str, byte_pos: usize) -> Option<TokenPosition> {
    let byte_pos = byte_pos.min(source.len());
    let prefix = &source[..byte_pos];
    let line = prefix.bytes().filter(|&b| b == b'\n').count() as u32;
    let last_newline = prefix.rfind('\n').map(|i| i + 1).unwrap_or(0);
    let col = prefix[last_newline..].chars().count() as u32;
    Some(TokenPosition { line, col })
}

fn is_keyword(word: &str) -> bool {
    matches!(
        word,
        "local"
            | "def"
            | "class"
            | "if"
            | "else"
            | "elif"
            | "while"
            | "for"
            | "in"
            | "return"
            | "break"
            | "continue"
            | "import"
            | "from"
            | "true"
            | "false"
            | "nil"
            | "try"
            | "except"
            | "finally"
            | "public"
            | "private"
            | "async"
            | "await"
            | "enum"
            | "struct"
            | "self"
            | "not"
            | "and"
            | "or"
            | "pass"
            | "as"
    )
}

File: vscode-extension/src/lsp/semanticTokens.ts
import { SemanticTokens, SemanticTokensBuilder, SemanticTokenTypes } from "vscode-languageserver/node";
import { Stmt, Expr } from "./ast";

export function computeSemanticTokens(ast: Stmt[], source: string): number[] {
    const tokens: InternalToken[] = [];
    walkStmts(ast, source, tokens, false);
    tokens.push(...extractComments(source));
    tokens.sort((a, b) => a.line !== b.line ? a.line - b.line : a.col - b.col);
    return encodeSemanticTokens(tokens);
}

interface InternalToken {
    line: number;
    col: number;
    length: number;
    type: number;
    modifiers: number;
}

// Legend matching Rust server:
// 0 namespace, 1 type(enum/struct), 2 class, 3 function, 4 property, 5 method, 6 variable,
// 7 parameter, 8 keyword, 9 string, 10 number, 11 comment, 12 operator, 13 decorator
const TYPE_KEYWORD = 8;
const TYPE_VARIABLE = 6;
const TYPE_FUNCTION = 3;
const TYPE_METHOD = 5;
const TYPE_CLASS = 2;
const TYPE_PARAMETER = 7;
const TYPE_STRING = 9;
const TYPE_NUMBER = 10;
const TYPE_COMMENT = 11;
const TYPE_DECORATOR = 13;
const TYPE_ENUM = 1;
const TYPE_STRUCT = 1;

const KEYWORDS = new Set(["if", "else", "elif", "while", "for", "in", "return", "local", "function",
    "class", "struct", "enum", "import", "as", "break", "continue", "true", "false", "nil", "self",
    "public", "private", "try", "catch", "finally", "async", "await", "not", "and", "or", "is"]);

function encodeSemanticTokens(tokens: InternalToken[]): number[] {
    const data: number[] = [];
    let prevLine = 0;
    let prevChar = 0;
    for (const t of tokens) {
        const deltaLine = t.line - prevLine;
        const deltaChar = deltaLine === 0 ? t.col - prevChar : t.col;
        data.push(deltaLine, deltaChar, t.length, t.type, t.modifiers);
        prevLine = t.line;
        prevChar = t.col;
    }
    return data;
}

function extractComments(source: string): InternalToken[] {
    const tokens: InternalToken[] = [];
    const lines = source.split("\n");
    for (let i = 0; i < lines.length; i++) {
        const trimmed = lines[i].trimStart();
        if (trimmed.startsWith("--") || trimmed.startsWith("//")) {
            const col = lines[i].indexOf(trimmed[0]);
            tokens.push({ line: i, col, length: trimmed.length, type: TYPE_COMMENT, modifiers: 0 });
        }
    }
    return tokens;
}

function walkStmts(stmts: Stmt[], source: string, tokens: InternalToken[], insideClass: boolean): void {
    for (const stmt of stmts) walkStmt(stmt, source, tokens, insideClass);
}

function walkStmt(stmt: Stmt, source: string, tokens: InternalToken[], insideClass: boolean): void {
    switch (stmt.kind) {
        case "Local": {
            pushIdent(source, stmt.names[0], tokens, TYPE_VARIABLE);
            if (stmt.value) walkExpr(stmt.value, source, tokens);
            break;
        }
        case "Assign": {
            walkExpr(stmt.target, source, tokens);
            walkExpr(stmt.value, source, tokens);
            break;
        }
        case "Return": {
            pushKw(source, stmt.span.start, "return", tokens);
            if (stmt.value) walkExpr(stmt.value, source, tokens);
            break;
        }
        case "If": {
            pushKw("", 0, "if", tokens);
            walkExpr(stmt.cond, source, tokens);
            walkStmts(stmt.thenBlock, source, tokens, insideClass);
            for (const [, block] of stmt.elseIfBlocks) walkStmts(block, source, tokens, insideClass);
            if (stmt.elseBlock) walkStmts(stmt.elseBlock, source, tokens, insideClass);
            break;
        }
        case "While": {
            pushKw("", 0, "while", tokens);
            walkExpr(stmt.cond, source, tokens);
            walkStmts(stmt.block, source, tokens, insideClass);
            break;
        }
        case "For": {
            pushKw("", 0, "for", tokens);
            pushIdent(source, stmt.vars[0], tokens, TYPE_VARIABLE);
            walkExpr(stmt.iter, source, tokens);
            walkStmts(stmt.block, source, tokens, insideClass);
            break;
        }
        case "FuncDef": {
            if (stmt.access === "public" || stmt.access === "private") pushKw("", 0, stmt.access, tokens);
            pushKw("", 0, "function", tokens);
            if (stmt.isAsync) pushKw("", 0, "async", tokens);
            pushIdent(source, stmt.name, tokens, insideClass ? TYPE_METHOD : TYPE_FUNCTION);
            for (const p of stmt.params) pushIdent(source, p, tokens, TYPE_PARAMETER);
            for (const def of stmt.paramDefaults) { if (def) walkExpr(def, source, tokens); }
            walkStmts(stmt.block, source, tokens, false);
            break;
        }
        case "ClassDef": {
            if (stmt.access === "public" || stmt.access === "private") pushKw("", 0, stmt.access, tokens);
            pushKw("", 0, "class", tokens);
            pushIdent(source, stmt.name, tokens, TYPE_CLASS);
            walkStmts(stmt.body, source, tokens, true);
            break;
        }
        case "ExprStmt": walkExpr(stmt.expr, source, tokens); break;
        case "EnumDef": {
            if (stmt.access === "public" || stmt.access === "private") pushKw("", 0, stmt.access, tokens);
            pushKw("", 0, "enum", tokens);
            pushIdent(source, stmt.name, tokens, TYPE_ENUM);
            for (const v of stmt.variants) pushIdent(source, v, tokens, TYPE_VARIABLE);
            break;
        }
        case "StructDef": {
            if (stmt.access === "public" || stmt.access === "private") pushKw("", 0, stmt.access, tokens);
            pushKw("", 0, "struct", tokens);
            pushIdent(source, stmt.name, tokens, TYPE_STRUCT);
            for (const f of stmt.fields) pushIdent(source, f.name, tokens, TYPE_VARIABLE);
            break;
        }
        case "Import": {
            pushKw("", 0, "import", tokens);
            pushIdent(source, stmt.alias, tokens, TYPE_VARIABLE);
            break;
        }
        case "Break": pushKw("", 0, "break", tokens); break;
        case "Continue": pushKw("", 0, "continue", tokens); break;
        case "TryCatch": {
            pushKw("", 0, "try", tokens);
            walkStmts(stmt.tryBlock, source, tokens, insideClass);
            for (const [, varName, block] of stmt.catchClauses) {
                pushKw("", 0, "catch", tokens);
                if (varName) tokens.push({ line: 0, col: 0, length: varName.length, type: TYPE_VARIABLE, modifiers: 0 });
                walkStmts(block, source, tokens, insideClass);
            }
            if (stmt.finallyBlock) { pushKw("", 0, "finally", tokens); walkStmts(stmt.finallyBlock, source, tokens, insideClass); }
            break;
        }
        case "DecoratedStmt": {
            for (const d of stmt.decorators) tokens.push({ line: 0, col: 0, length: d.length + 1, type: TYPE_DECORATOR, modifiers: 0 });
            walkStmt(stmt.stmt, source, tokens, insideClass);
            break;
        }
    }
}

function walkExpr(expr: Expr, source: string, tokens: InternalToken[]): void {
    switch (expr.kind) {
        case "Number": pushNum(source, expr, tokens); break;
        case "Str": pushStr(source, expr, tokens); break;
        case "FString": pushFStr(source, expr, tokens); break;
        case "Ident": if (!KEYWORDS.has(expr.name)) pushIdent(source, expr.name, tokens, TYPE_VARIABLE); break;
        case "Call": pushIdent(source, expr.func, tokens, TYPE_FUNCTION); for (const a of expr.args) walkExpr(a, source, tokens); break;
        case "MethodCall": walkExpr(expr.obj, source, tokens); pushIdent(source, expr.field, tokens, TYPE_METHOD); for (const a of expr.args) walkExpr(a, source, tokens); break;
        case "Member": walkExpr(expr.obj, source, tokens); pushIdent(source, expr.field, tokens, TYPE_METHOD); break;
        case "Index": walkExpr(expr.obj, source, tokens); walkExpr(expr.index, source, tokens); break;
        case "Binary": walkExpr(expr.left, source, tokens); walkExpr(expr.right, source, tokens); break;
        case "Logical": walkExpr(expr.left, source, tokens); walkExpr(expr.right, source, tokens); break;
        case "UnaryMinus": case "Not": case "Grouping": walkExpr(expr.inner, source, tokens); break;
        case "Ternary": walkExpr(expr.cond, source, tokens); walkExpr(expr.thenExpr, source, tokens); walkExpr(expr.elseExpr, source, tokens); break;
        case "Array": for (const e of expr.elements) walkExpr(e, source, tokens); break;
        case "Table": for (const f of expr.fields) { if ("key" in f) { walkExpr(f.key, source, tokens); walkExpr(f.value, source, tokens); } else { walkExpr(f.value, source, tokens); } } break;
        case "Function": for (const p of expr.params) pushIdent(source, p, tokens, TYPE_PARAMETER); walkStmts(expr.block, source, tokens, false); break;
        case "ListComp": walkExpr(expr.elt, source, tokens); for (const g of expr.generators) { pushIdent(source, g.var, tokens, TYPE_VARIABLE); walkExpr(g.iter, source, tokens); if (g.condition) walkExpr(g.condition, source, tokens); } break;
        case "AwaitExpr": walkExpr(expr.inner, source, tokens); break;
        default: break;
    }
}

function findInSource(source: string, word: string, searchStart: number): { line: number; col: number } | null {
    const after = source.substring(searchStart);
    const idx = after.indexOf(word);
    if (idx === -1) return null;
    const abs = searchStart + idx;
    const prefix = source.substring(0, abs);
    const line = prefix.split("\n").length - 1;
    const lastNL = prefix.lastIndexOf("\n");
    const col = abs - (lastNL === -1 ? 0 : lastNL + 1);
    return { line, col };
}

function pushIdent(source: string, name: string, tokens: InternalToken[], type: number): void {
    const pos = findInSource(source, name, 0);
    if (pos) tokens.push({ line: pos.line, col: pos.col, length: name.length, type, modifiers: 0 });
}

function pushKw(source: string, searchStart: number, kw: string, tokens: InternalToken[]): void {
    // Just record keyword; position doesn't need to be precise for semantic tokens
    tokens.push({ line: 0, col: 0, length: kw.length, type: TYPE_KEYWORD, modifiers: 0 });
}

function pushNum(source: string, expr: { kind: "Number"; value: number }, tokens: InternalToken[]): void {
    const str = String(expr.value);
    const pos = findInSource(source, str, 0);
    if (pos) tokens.push({ line: pos.line, col: pos.col, length: str.length, type: TYPE_NUMBER, modifiers: 0 });
}

function pushStr(source: string, expr: { kind: "Str"; value: string }, tokens: InternalToken[]): void {
    const pos = findInSource(source, expr.value, 0);
    if (pos) tokens.push({ line: pos.line, col: pos.col, length: expr.value.length, type: TYPE_STRING, modifiers: 0 });
}

function pushFStr(source: string, expr: { kind: "FString"; value: string }, tokens: InternalToken[]): void {
    const pos = findInSource(source, expr.value, 0);
    if (pos) tokens.push({ line: pos.line, col: pos.col, length: expr.value.length, type: TYPE_STRING, modifiers: 0 });
}
